src/ga4gh/cat_vrs/recipes.py

```diff
@@ -0,0 +1,184 @@
+    that is representative of a collection of congruent Protein Alleles that share the
+    same altered codon(s).
+    """
+
+    constraints: list[Constraint] = Field(..., min_length=1)
+
+    @field_validator("constraints")
+    @classmethod
+    def validate_constraints(cls, v: list[Constraint]) -> list[Constraint]:
+        """Validate constraints property
+
+        At least one constraint in ``constraints`` must satisfy ALL of the following
+        requirements:
+        1. Must be a ``DefiningAlleleConstraint``
+        2. Must have ``relations`` property that meets ALL of the following
+        requirements:
+            a. Must contain exactly one relation where ``primaryCode = translates_from``
+
+        :param v: Constraints property to validate
+        :raises ValueError: If constraints property does not satisfy the requirements
+        :return: Constraints property
+        """
+        if not any(
+            isinstance(constraint.root, DefiningAlleleConstraint)
+            and constraint.root.relations
+            and sum(
+                1
+                for r in constraint.root.relations
+                if r.primaryCode
+                and r.primaryCode.root == Relation.TRANSLATES_FROM.value
+            )
+            == 1
+            for constraint in v
+        ):
+            err_msg = f"Unable to find at least one constraint that is a `DefiningAlleleConstraint` and has exactly one `relation` where the `primaryCode` is '{Relation.TRANSLATES_FROM.value}'."
+            raise ValueError(err_msg)
+
+        return v
+
+
+class CanonicalAllele(CategoricalVariant):
+    """A canonical allele is defined by an
+    `Allele <https://vrs.ga4gh.org/en/2.x/concepts/MolecularVariation/Allele.html#>`_
+    that is representative of a collection of congruent Alleles, each of which depict
+    the same nucleic acid change on different underlying reference sequences. Congruent
+    representations of an Allele often exist across different genome assemblies and
+    associated cDNA transcript representations.
+    """
+
+    constraints: list[Constraint] = Field(..., min_length=1, max_length=1)
+
+    @field_validator("constraints")
+    @classmethod
+    def validate_constraints(cls, v: list[Constraint]) -> list[Constraint]:
+        """Validate constraints property
+
+        Exactly one constraint in ``constraints`` must satisfy ALL of the following
+        requirements:
+        1. Must be a ``DefiningAlleleConstraint``
+        2. Must have ``relations`` property that meets ALL of the following
+        requirements:
+            a. Must contain exactly one relation where ``primaryCode = liftover_to``
+            b. Must contain exactly one relation where ``primaryCode = transcribes_to``
+
+        :param v: Constraints property to validate
+        :raises ValueError: If constraints property does not satisfy the requirements
+        :return: Constraints property
+        """
+        constraint = v[0]
```

Comment:
Here, this presumes that the list of constraints is of length 1. Should there be a similar "if not" statement checking the length of the `list[Constraint]` input? e.g.,

```python
if not len(v) == 1:
	err_msg = "CanonicalAllele requires a minimum and maximum of 1 constraint"
	raise ValueError(err_msg)
```

Without this error check, users could pass a list of constraints for a canonical allele and this function will only check the first constraint. 

With that said, this will _probably_ be changed in a future branch per adding FeatureContextConstraint.

Reply:
@brendanreardon So in line 74 we have: `constraints: list[Constraint] = Field(..., min_length=1, max_length=1)` to ensure this list has a length of 1